notepad-plus-plus/notepad-plus-plus
# prevent reloading two incompatible copies of same file, fix issue #10266

Fix issue #10266, concerning unsaved cloned documents being reloaded.

When a buffer's status is saved in config.xml, the file manager checks if it is open in both edit views. If so, a hasClone attribute for that file's session.xml entry is set to `yes`.

When the session is loaded, all buffers in the main edit view are loaded as normal, and then when buffers in the subeditview are loaded, any that have the hasClone attribute check for a same-name file in the main edit view and clone it into the subeditview.

TO TEST:
1. Create a `new 1` document in one edit view, then make an edit.
2. Clone `new 1` into another view.
3. Open a real file (not a `new X` file) and clone it into another view, then edit one of the clones. DO NOT SAVE THE FILE.
4. Close the session.
5. Reopen the session.
6. Try editing the `new 1` document that was loaded from session. These edits should be mirrored in the copy in the other edit view.
7. Try editing the real file, and verify that it too is cloned, not split into two incompatible versions.

__CAVEAT:__
Arguably it would be preferable for Notepad++ to assume that *any time a document with the same name is open in both edit views, those two copies are clones.* The solution I have chosen *does not do that*, which means that if you clone a document in an older version of Notepad++, dirty the clones, and reopen those files in a version of Notepad++ with this fix, those will still be treated as two incompatible copies of the same document.

Presumably an alternate way to solve the problem would be to do something about the way that `Notepad_plus::loadSession` decides whether to clone a document in the subEditView when loading from session.xml. As it stands, it seems that this process involves checking to see if the buffer is dirty, and maybe that check should be eliminated.
